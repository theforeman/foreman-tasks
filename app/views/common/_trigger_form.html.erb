<div class="form-group ">
  <label class="col-md-2 control-label"><%= _('Schedule') %></label>
  <div class="col-md-6">
    <% javascript 'foreman_tasks/foreman_tasks' %>
    <% stylesheet 'foreman_tasks/foreman_tasks' %>

    <%= javascript_tag do %>
      $(function() { trigger_form_selector_binds('<%= f.options[:html][:id] %>','<%= f.object_name %>') });
    <% end %>

    <%= fields_for :triggering, triggering do |trigger_fields| %>
      <%= radio_button_f trigger_fields, :mode, :class => 'trigger_mode_selector', :value => 'immediate', :text => _("Execute now") %>
      <%= radio_button_f trigger_fields, :mode, :class => 'trigger_mode_selector', :value => 'future',    :text => _("Schedule future execution") %>
      <%= radio_button_f trigger_fields, :mode, :class => 'trigger_mode_selector', :value => 'recurring', :text => _("Set up recurring execution") %>
  </div>
</div>

<div class="trigger_fields">
  <%= future_mode_fieldset trigger_fields, triggering %>
  <%= recurring_mode_fieldset trigger_fields, triggering %>
</div>
<% end %>

