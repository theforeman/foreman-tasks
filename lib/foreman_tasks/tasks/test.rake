namespace :test do
  task :foreman_tasks => 'db:test:prepare' do
    test_task = Rake::TestTask.new('foreman_tasks_test_task') do |t|
      t.libs << ['test', "#{ForemanTasks::Engine.root}/test"]
      t.test_files = ["#{ForemanTasks::Engine.root}/test/**/*_test.rb"]
      t.verbose = true
      t.warning = false
    end

    Rake::Task[test_task.name].invoke
  end
end

namespace :foreman_tasks do
  task :rubocop do
    begin
      require 'rubocop/rake_task'
      RuboCop::RakeTask.new(:rubocop_foreman_tasks) do |task|
	task.patterns = ["#{ForemanTasks::Engine.root}/app/**/*.rb",
		  "#{ForemanTasks::Engine.root}/lib/**/*.rb",
		  "#{ForemanTasks::Engine.root}/test/**/*.rb"]
      end
    rescue
      puts 'Rubocop not loaded.'
    end

    Rake::Task['rubocop_foreman_tasks'].invoke
  end
end

Rake::Task[:test].enhance do
  Rake::Task['test:foreman_tasks'].invoke
end

load 'tasks/jenkins.rake'
if Rake::Task.task_defined?(:'jenkins:unit')
  Rake::Task['jenkins:unit'].enhance ['foreman_tasks:rubocop', 'test:foreman_tasks']
end
