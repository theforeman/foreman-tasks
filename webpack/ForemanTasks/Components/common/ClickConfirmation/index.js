import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  Modal,
  ModalVariant,
  Checkbox,
  TextContent,
} from '@patternfly/react-core';
import { translate as __, sprintf } from 'foremanReact/common/I18n';
import {
  UNLOCK_MODAL,
  FORCE_UNLOCK_MODAL,
} from '../../TaskActions/TaskActionsConstants';

export const ClickConfirmation = ({
  title,
  confirmType,
  body,
  confirmationMessage,
  id,
  confirmAction,
  onClick,
  isOpen,
  setModalClosed,
}) => {
  const [isConfirmed, setConfirm] = useState(false);
  const onClose = () => {
    setConfirm(false);
    setModalClosed();
  };
  return (
    <Modal
      title={title}
      titleIconVariant={confirmType}
      variant={ModalVariant.small}
      isOpen={isOpen}
      onClose={onClose}
      ouiaId={`${id}-modal`}
      actions={[
        <Button
          ouiaId={`${id}-cancel-button`}
          key="cancel"
          variant="secondary"
          onClick={onClose}
        >
          {__('Cancel')}
        </Button>,
        <Button
          ouiaId={`${id}-confirm-button`}
          key="confirm"
          className="confirm-button"
          variant={confirmType === 'danger' ? 'danger' : 'primary'}
          onClick={() => {
            onClick();
            onClose();
          }}
          isDisabled={!isConfirmed}
        >
          {confirmAction}
        </Button>,
      ]}
    >
      <TextContent>
        {body}
        <Checkbox
          ouiaId={`${id}-confirmation-checkbox`}
          id={`confirmation-checkbox-${id}`}
          label={confirmationMessage}
          isChecked={isConfirmed}
          onChange={(_e, checked) => setConfirm(checked)}
        />
      </TextContent>
    </Modal>
  );
};

ClickConfirmation.propTypes = {
  title: PropTypes.string.isRequired,
  body: PropTypes.string.isRequired,
  confirmationMessage: PropTypes.string.isRequired,
  confirmAction: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  confirmType: PropTypes.oneOf(['warning', 'danger']),
  id: PropTypes.string.isRequired,
  isOpen: PropTypes.bool,
  setModalClosed: PropTypes.func,
};

ClickConfirmation.defaultProps = {
  confirmType: 'warning',
  isOpen: false,
  setModalClosed: () => {},
};

// Unlock Modal Components
const confirmationMessage = __(
  'I understand that this may cause harm and have working database backups of all backend services.'
);

export const UnlockModal = ({ onClick, id, isOpen, setModalClosed }) => (
  <ClickConfirmation
    id={id}
    title={__('Unlock')}
    body={__(
      "This will unlock the resources that the task is running against. Please note that this might lead to inconsistent state and should be used with caution, after making sure that the task can't be resumed."
    )}
    confirmationMessage={confirmationMessage}
    confirmAction={__('Unlock')}
    onClick={onClick}
    confirmType="warning"
    isOpen={isOpen}
    setModalClosed={setModalClosed}
  />
);

export const ForceUnlockModal = ({
  onClick,
  id,
  selectedRowsLen,
  isOpen,
  setModalClosed,
}) => (
  <ClickConfirmation
    id={id}
    title={__('Force Unlock')}
    body={sprintf(
      __(
        `Resources for %s task(s) will be unlocked and will not prevent other tasks from being run. As the task(s) might be still running, it should be avoided to use this unless you are really sure the task(s) got stuck.`
      ),
      selectedRowsLen
    )}
    confirmationMessage={confirmationMessage}
    confirmAction={__('Force Unlock')}
    onClick={onClick}
    confirmType="danger"
    isOpen={isOpen}
    setModalClosed={setModalClosed}
  />
);

UnlockModal.propTypes = {
  onClick: PropTypes.func.isRequired,
  id: PropTypes.string,
  isOpen: PropTypes.bool,
  setModalClosed: PropTypes.func,
};

ForceUnlockModal.propTypes = {
  onClick: PropTypes.func.isRequired,
  id: PropTypes.string,
  selectedRowsLen: PropTypes.number,
  isOpen: PropTypes.bool,
  setModalClosed: PropTypes.func,
};

UnlockModal.defaultProps = {
  id: UNLOCK_MODAL,
  isOpen: false,
  setModalClosed: () => {},
};

ForceUnlockModal.defaultProps = {
  id: FORCE_UNLOCK_MODAL,
  selectedRowsLen: 1,
  isOpen: false,
  setModalClosed: () => {},
};

export default ClickConfirmation;
