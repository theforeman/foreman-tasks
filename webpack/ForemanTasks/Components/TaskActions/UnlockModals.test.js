import { testComponentSnapshotsWithFixtures } from '@theforeman/test';

import { UnlockModal, ForceUnlockModal } from '../common/ClickConfirmation';

const fixtures = {
  render: { taskID: 'some-id', onClick: jest.fn() },
};

describe('UnlockModal', () => {
  testComponentSnapshotsWithFixtures(UnlockModal, fixtures);
});
describe('ForceUnlockModal', () => {
  testComponentSnapshotsWithFixtures(ForceUnlockModal, fixtures);
});
