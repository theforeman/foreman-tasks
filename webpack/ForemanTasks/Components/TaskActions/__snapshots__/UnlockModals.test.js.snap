// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForceUnlockModal render 1`] = `
<ClickConfirmation
  body="Resources for 1 task(s) will be unlocked and will not prevent other tasks from being run. As the task(s) might be still running, it should be avoided to use this unless you are really sure the task(s) got stuck."
  confirmAction="Force Unlock"
  confirmType="danger"
  confirmationMessage="I understand that this may cause harm and have working database backups of all backend services."
  id="forceUnlockModal"
  isOpen={false}
  onClick={[MockFunction]}
  setModalClosed={[Function]}
  title="Force Unlock"
/>
`;

exports[`UnlockModal render 1`] = `
<ClickConfirmation
  body="This will unlock the resources that the task is running against. Please note that this might lead to inconsistent state and should be used with caution, after making sure that the task can't be resumed."
  confirmAction="Unlock"
  confirmType="warning"
  confirmationMessage="I understand that this may cause harm and have working database backups of all backend services."
  id="unlockModal"
  isOpen={false}
  onClick={[MockFunction]}
  setModalClosed={[Function]}
  title="Unlock"
/>
`;
